Add compatibility notice to sidebar when editing Cart and Checkout blocks (and their inner blocks)

opr opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR adds a notice to the inspector when editing Cart and Checkout blocks, and their inner blocks. It also adds functionality to dismiss the notice independently on Cart and Checkout blocks.

I reused the `useCompatibilityNotice` functionality we have, and used the block names. We don't need to make new ones since `cart` and `checkout` were currently being used to track the old pop-up compatibility notice. This won't affect mini cart as that could use its own key to track its own compatibility notice status.

<!-- Reference any related issues or PRs here -->

Fixes #6851

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

<img width="293" alt="image" src="https://user-images.githubusercontent.com/5656702/182901522-cc520aa8-e203-45c5-9aff-de443ee0efbe.png">

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Before checking this PR out, build `trunk`, clear local storage and visit the Cart/Checkout blocks. Dismiss the compatibility notices on each page. Then check out this PR.
2. Go to the editor, insert the Checkout block.
3. Select it, verify you see the notice. Click through each and every inner block and ensure the notice displays for it.
4. Dismiss the notice, click back through the inner blocks and ensure the notice is gone. Reload the page, ensure the notice is still gone.
5. Go to the editor, insert the Cart block.
6. Select it, verify you see the notice. Click through each and every inner block and ensure the notice displays for it.
7. Dismiss the notice, click back through the inner blocks and ensure the notice is gone. Reload the page, ensure the notice is still gone.
8. Clear local storage again.
9. Go back to the Cart and Checkout blocks and ensure the notice is back.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Add notice to Cart and Checkout blocks' inspector controls which links to the list of compatible plugins.

opr committed:
Add CartCheckoutSidebarCompatibilityNotice component and styles

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in Checkout main block

opr committed:
Add sidebarCompatabilityControls function

opr committed:
Remove sidebarCompatibilityControls func, replace with component

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in express payments block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in order notes block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in order summary block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in order summary items

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in checkout coupon block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in discount block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in checkout fees block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in shipping block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in subtotal block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in checkout taxes block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in checkout payment block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in shipping block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in checkout terms block

opr committed:
Include CartCheckoutSidebarCompatibilityNotice in checkout totals block

opr committed:
Get if notice is dismissed from useCompatibilityNotice

opr committed:
Export one component, one for each Cart and Checkout

opr committed:
Add styling for top border of notice

opr committed:
Rename to CheckoutSidebarCompatibilityNotice

opr committed:
Add CartSidebarCompatibilityNotice to accepted payment methods block

opr committed:
Add CartSidebarCompatibilityNotice to express payment methods block

opr committed:
Add CartSidebarCompatibilityNotice to cart items block

opr committed:
Add CartSidebarCompatibilityNotice to cart line items block

opr committed:
Add CartSidebarCompatibilityNotice to cart order summary block

opr committed:
Add CartSidebarCompatibilityNotice to cart coupons block

opr committed:
Add CartSidebarCompatibilityNotice to cart discounts block

opr committed:
Add CartSidebarCompatibilityNotice to cart fees block

opr committed:
Add CartSidebarCompatibilityNotice to cart summary heading block

opr committed:
Add CartSidebarCompatibilityNotice to cart shipping block

opr committed:
Add CartSidebarCompatibilityNotice to cart subtotal block

opr committed:
Move import to correct location

opr committed:
Add CartSidebarCompatibilityNotice to cart taxes block

opr committed:
Add CartSidebarCompatibilityNotice to cart totals block

opr committed:
Add CartSidebarCompatibilityNotice to empty cart block

opr committed:
Wrap notice in div to correctly style with border and padding

opr committed:
Add CartSidebarCompatibilityNotice to filled cart block

opr committed:
Add CartSidebarCompatibilityNotice to cart proceed to checkout block

opr committed:
Don't use /inspector suffix for compatibility tracking

opr committed:
Remove old compatibility notice from cart and checkout blocks

tarhi-saad commented on assets/js/blocks/cart/edit.js:
```diff
@@ -38,6 +38,7 @@ import {
 	useForcedLayout,
 } from '../cart-checkout-shared';
 import { CartBlockContext } from './context';
+import { CartSidebarCompatibilityNotice } from '@woocommerce/editor-components/sidebar-compatibility-notice';
```
We should move this new notice component to the `External dependencies` list. And remove the old one in `line 15`. 🙌

tarhi-saad commented on assets/js/blocks/cart/edit.js:
```diff
@@ -111,6 +112,7 @@ const BlockSettings = ( { attributes, setAttributes } ) => {
 				/>
 			</PanelBody>
 			<CartCheckoutFeedbackPrompt />
+			<CartSidebarCompatibilityNotice />
```
We should place the alert under the `Name & description` panel before the Block's settings panels (i.e., Let's move it to `line 70`).

tarhi-saad commented on assets/js/blocks/cart/inner-blocks/cart-order-summary-taxes/edit.tsx:
```diff
@@ -63,6 +64,7 @@ export const Edit = ( {
 							/>
 						</PanelBody>
 					) }
+				<CartSidebarCompatibilityNotice />
```
Same here! We should place the alert under the `Name & description` panel before the Block's settings panels (i.e., Let's move it to `line 39`).

tarhi-saad commented on assets/js/blocks/cart/inner-blocks/cart-order-summary-shipping/edit.tsx:
```diff
@@ -59,6 +60,7 @@ export const Edit = ( {
 						/>
 					</PanelBody>
 				) }
+				<CartSidebarCompatibilityNotice />
```
Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 37).

tarhi-saad commented on assets/js/blocks/cart/inner-blocks/proceed-to-checkout-block/edit.tsx:
```diff
@@ -59,6 +62,7 @@ export const Edit = ( {
 						} }
 					/>
 				) }
+				<CartSidebarCompatibilityNotice />
```
Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 45).

tarhi-saad commented on assets/js/blocks/checkout/edit.tsx:
```diff
@@ -29,6 +29,8 @@ import { CHECKOUT_PAGE_ID } from '@woocommerce/block-settings';
 import { getAdminLink } from '@woocommerce/settings';
 import { CartCheckoutCompatibilityNotice } from '@woocommerce/editor-components/compatibility-notices';
+import { CheckoutSidebarCompatibilityNotice } from '@woocommerce/editor-components/sidebar-compatibility-notice';
+import {} from '@woocommerce/editor-components/compatibility-notices';
```
We should remove the imports at `line 33` & `line 31`!

tarhi-saad commented on assets/js/blocks/checkout/inner-blocks/checkout-order-summary-taxes/edit.tsx:
```diff
@@ -63,6 +64,7 @@ export const Edit = ( {
 							/>
 						</PanelBody>
 					) }
+				<CheckoutSidebarCompatibilityNotice />
```
Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 39).

tarhi-saad commented on assets/js/blocks/checkout/inner-blocks/checkout-payment-block/edit.tsx:
```diff
@@ -85,6 +86,7 @@ export const Edit = ( {
 						</ExternalLink>
 					</PanelBody>
 				) }
+				<CheckoutSidebarCompatibilityNotice />
```
Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 56).

tarhi-saad commented on assets/js/blocks/checkout/inner-blocks/checkout-shipping-methods-block/edit.tsx:
```diff
@@ -119,6 +120,7 @@ export const Edit = ( {
 						</ExternalLink>
 					</PanelBody>
 				) }
+				<CheckoutSidebarCompatibilityNotice />
```
Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 60).

tarhi-saad commented on assets/js/blocks/checkout/inner-blocks/checkout-terms-block/edit.tsx:
```diff
@@ -55,6 +56,7 @@ export const Edit = ( {
 						}
 					/>
 				</PanelBody>
+				<CheckoutSidebarCompatibilityNotice />
```
Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 40).

tarhi-saad commented on assets/js/blocks/checkout/edit.tsx:
```diff
@@ -109,6 +111,7 @@ const BlockSettings = ( {
 				/>
 			</PanelBody>
 			<CartCheckoutFeedbackPrompt />
+			<CheckoutSidebarCompatibilityNotice />
```
Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 69).

tarhi-saad commented on assets/js/blocks/checkout/edit.tsx:
```diff
@@ -166,6 +169,7 @@ export const Edit = ( {
 					}
 				/>
 			</PanelBody>
+			<CheckoutSidebarCompatibilityNotice />
```
Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 153).

tarhi-saad commented on assets/js/blocks/checkout/edit.tsx:
```diff
@@ -225,6 +229,7 @@ export const Edit = ( {
 					/>
 				) }
 			</PanelBody>
+			<CheckoutSidebarCompatibilityNotice />
```
Same here! We should place the alert under the Name & description panel before the Block's settings panels (i.e., Let's move it to line 178).

tarhi-saad commented on assets/js/editor-components/sidebar-compatibility-notice/style.scss:
```diff
@@ -0,0 +1,13 @@
+.wc-blocks-sidebar-compatibility-notice__notice.components-notice.is-dismissible {
+	margin: 0;
+	padding: $gap;
+	padding-right: $gap-small;
```
We have more `padding` to the right which produces an uneven spacing.

<img width="306" alt="image" src="https://user-images.githubusercontent.com/14235870/183070154-50fa063f-58cf-42ce-959c-80b013abd5e7.png">

tarhi-saad commented on assets/js/editor-components/sidebar-compatibility-notice/style.scss:
```diff
@@ -0,0 +1,13 @@
+	padding-right: $gap-small;
+
+	.components-notice__content {
+		margin-right: 0;
```
This style isn't applied on my end!

<img width="233" alt="image" src="https://user-images.githubusercontent.com/14235870/183071019-0d166dde-d5ee-40b6-beab-c47f0ae4622a.png">

tarhi-saad commented on assets/js/editor-components/sidebar-compatibility-notice/style.scss:
```diff
@@ -0,0 +1,13 @@
+.wc-blocks-sidebar-compatibility-notice {
+	border-top: 1px solid $gray-200;
```
There is no `border-top` from my side!

<img width="297" alt="image" src="https://user-images.githubusercontent.com/14235870/183071172-ccbef80e-a192-483e-8948-bd4486435278.png">

tarhi-saad commented on assets/js/editor-components/sidebar-compatibility-notice/index.tsx:
```diff
@@ -0,0 +1,51 @@
+			a: (
+				// Suppress the warning as this <a> will be interpolated into the string with content.
+				// eslint-disable-next-line jsx-a11y/anchor-has-content
+				<ExternalLink href="https://woocommerce.com/document/cart-checkout-blocks-support-status/#section-7" />
```
We should redirect to the list of `Compatible Extensions` (i.e., this link instead: https://woocommerce.com/document/cart-checkout-blocks-support-status/#section-3). In the issue description that @ralucaStan wrote #6851, we are mentioning the list of `Incompatible Extensions` instead. But, on the Figma file and on the context post pdFofs-b7-p2, we want to redirect to the `Compatible Extensions`.

tarhi-saad reviewed (changes requested):
Thank you, @opr, for working on this PR! 🙌  I've left some comments below! 👍 I apologize for leaving a lot of comments regarding the same issue: 

```
The alert will:

be placed under the Name & description panel, before the Block's settings panels
``` 
I just wanted to make it easy to find all the files and lines that should be updated! 🙌

opr committed:
Use editor style instead of style.scss

opr committed:
Move CartSidebarCompatibilityNotice to correct position

opr committed:
Move import to correct location and remove unused one

opr committed:
Remove unused imports

opr committed:
Add prop to Controls component to allow showing/hiding compat notice

opr committed:
Include CheckoutSidebarCompatibilityNotice in FormStep

opr committed:
Include CheckoutSidebarCompatibilityNotice in checkout actions block

opr committed:
Include CheckoutSidebarCompatibilityNotice in billing address

opr committed:
Remove CheckoutSidebarCompatibilityNotice from shipping methods block

opr committed:
Remove CheckoutSidebarCompatibilityNotice from payment methods block

opr committed:
Move CheckoutSidebarCompatibilityNotice to correct position

opr committed:
Move CheckoutSidebarCompatibilityNotice to correct place in taxes block

opr committed:
Update link to compatible extensions

opr commented:
Hey @tarhi-saad thanks a lot for the detailed review! I actually moved things about a bit more as sometimes it wasn't possible to position the notice at the top due to other controls being used.

I've added a `includeCompatibilityNotice` prop to the `addressFieldControls` to allow us to turn the notice off if required. This was needed for the `checkout fields` block since it was showing two notices. Trying to get it to only show one affected the address field blocks, and the contact information block.

I also changed the stylesheet to be called `editor.scss` so it'll be included in our editor stylesheet.

opr committed:
Remove unused imports

opr committed:
Move useCompatibilityNotice to sidebar-compatibility-notice directory

opr committed:
Remove old CartCheckoutCompatibilityNotice
